packages/name-controller/src/providers/etherscan.ts

```diff
@@ -0,0 +1,101 @@
+import { ETHERSCAN_SUPPORTED_NETWORKS } from '../constants';
+import type {
+  NameProvider,
+  NameProviderMetadata,
+  NameProviderRequest,
+  NameProviderResult,
+} from '../types';
+import { NameType } from '../types';
+import { handleFetch } from '../util';
+
+const ID = 'etherscan';
+const LABEL = 'Etherscan (Verified Contract Name)';
+
+type EtherscanGetSourceCodeResponse = {
+  status: '1' | '0';
+  message: string;
+  result: [
+    {
+      SourceCode: string;
+      ABI: string;
+      ContractName: string;
+      CompilerVersion: string;
+      OptimizationUsed: string;
+      Runs: string;
+      ConstructorArguments: string;
+      Library: string;
+      LicenseType: string;
+      Proxy: string;
+      Implementation: string;
+      SwarmSource: string;
+    },
+  ];
+};
+
+export class EtherscanNameProvider implements NameProvider {
+  #apiKey?: string;
+
+  constructor({ apiKey }: { apiKey?: string } = {}) {
+    this.#apiKey = apiKey;
+  }
+
+  getMetadata(): NameProviderMetadata {
+    return {
+      sourceIds: { [NameType.ETHEREUM_ADDRESS]: [ID] },
+      sourceLabels: { [ID]: LABEL },
+    };
+  }
+
+  async getProposedNames(
+    request: NameProviderRequest,
+  ): Promise<NameProviderResult> {
+    const { value, chainId } = request;
+
+    const url = this.#getUrl(chainId, {
+      module: 'contract',
+      action: 'getsourcecode',
+      address: value,
+      apikey: this.#apiKey,
+    });
+
+    const responseData = (await handleFetch(
```

Comment:
Is it worth giving `handleFetch` a type parameter so you don't have to typecast the result?

Reply:
For the sake of clarity and stability, I wanted those functions to match exactly their counterparts in `controller-utils` so it would be trivial to adopt `controller-utils` in future.

Reply:
Ah, good call!